### PR TITLE
FlexAttention support for NJT

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #138098
* __->__ #136792
* #138841

This PR adds FlexAttention + NJT support. In particular:
* To handle raggedness, treats the packed sequence dim of input NJTs as a giant "stacked sequence". To ensure user `score_mod` / `mask_mod` functions can still be written in the original NJT sequence space, this PR handles conversions for indices within the giant "stacked sequence" -> sequence relative indices automatically.
* Provides `py_impls` for `NestedTensor` to the HOPs for flex attention forward / backward that simply wrap / unwrap NJTs appropriately
* Adds barebones `new_empty()` support to NJT since FlexAttention utilizes this repeatedly; right now, only `new_empty()` with a shape of `()` is supported
* Tests that FlexAttention with a causal mask matches causal SDPA
* Adds a new public API for FlexAttention usage:
    * `create_nested_block_mask(mask_mod, B, H, njt, BLOCK_SIZE, _compile)` - NJT analogue for `create_block_mask()` that utilizes the `njt`'s ragged structure to create an appropriately-sized block mask (e.g. `(1, 1, total_seqlen, total_seqlen)`). This function handles the index conversion from "stacked sequence" space -> relative sequence space. 
      * Minor note: as this is a public API, this function is purposefully named with "nested" instead of "njt" to keep the latter as an informal, mostly internal-only term.

Example usage:
```python
def causal_mask(b, h, q_idx, kv_idx):
    return q_idx >= kv_idx

query = ... # NJT of shape (B, H, S*, D)
key = ... # NJT of shape (B, H, S*, D)
value = ... # NJT of shape (B, H, S*, D)
# create_nested_block_mask() automatically converts indices from "stacked sequence" space -> relative sequence space
block_mask = create_nested_block_mask(causal_mask, 1, 1, query)  # block mask conceptual shape is (B, H, sum(S*), sum(S*))
output = flex_attention(query, key, value, block_mask=block_mask)

def causal_score_mod(score, b, h, q_idx, kv_idx):
    return torch.where(q_idx >= kv_idx, score, float("-inf"))

# flex_attention() automatically converts indices from "stacked sequence" space -> relative sequence space for NJT inputs
output2 = flex_attention(query, key, value, score_mod=causal_score_mod)
```

TODO:
* ~~Determine the right level of abstraction for public API helpers + move them alongside other helpers~~ Verify this with others though
* ~~Some cleanup~~
* ~~`njt_score_mod_adapter`~~
* ~~Q: should `create_njt_block_mask()` call `njt_mask_mod_adapter()` so we don't need two calls?~~
* Can we avoid materializing the `sum(s)` length `seq_idx` used for conversion between stacked sequence -> sequence relative indices?
    * Not for now, although future work may deepen the integration between Flex + NJT (possibly requiring custom templates). We should try to cache this though.
* ~~Demonstrate non-causal mask~~
* Support non-contiguous NJTs with holes (**booted to future PR**)

cc @cpuhrsch @bhosmer @drisspg @soulitzer @davidberard98 @YuqingJ @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @rec

### CODE REVIEW

File: torch/nn/attention/flex_attention.py [hunk 1]
```diff
@@ -1030,11 +1144,13 @@ def score_mod(
...
             )
 
+    from torch.fx.experimental.symbolic_shapes import is_nested_int
+
     if score_mod is None:
         score_mod = _identity
     if block_mask is None:
         block_mask = _create_empty_block_mask(query, key)
-    elif (
+    elif not is_nested_int(query.size(-2)) and (
```
Comment (jbschlosser) [inline torch/nn/attention/flex_attention.py:1153]: I hacked around this but is this important to have for NJT as well ? The issue is that the logic at this level is before the HOP is called, and the HOP is where the NJT is unwrapped to pass the `values` buffer to flex. So `query` is still an NJT at this point, `query.size(-2)` is usually a nested int, and nested ints don't support the lt operator.

File: test/inductor/test_flex_attention.py [hunk 2]
```diff
@@ -32,18 +32,13 @@
...
 from torch.testing._internal import common_utils
 from torch.testing._internal.common_cuda import PLATFORM_SUPPORTS_BF16, TEST_MULTIGPU
+from torch.testing._internal.common_device_type import (
+    flex_attention_supported_platform as supported_platform,
+)
 from torch.testing._internal.common_utils import TEST_WITH_ROCM
 from torch.utils._triton import has_triton
 
 
-# Skip tests if Triton is not available
```
Comment (drisspg) [inline test/inductor/test_flex_attention.py:39]: bring back pleaseeeeee

File: test/inductor/test_flex_attention.py [hunk 3]
```diff
@@ -2372,6 +2367,23 @@ def forward(self, L_query_: "f64[2, 2, 128, 4]", L_key_: "f64[2, 2, 128, 4]", L_
...
+
+        getattr_6 = l_key_.is_nested;  getattr_6 = None
+
+        getattr_7 = l_value_.is_nested;  getattr_7 = None
+
+        getattr_8 = l_query_.is_nested;  getattr_8 = None
+
+        getattr_9 = l_query_.is_nested;  getattr_9 = None
+
+        getattr_10 = l_query_.is_nested;  getattr_10 = None
```
Comment (jbschlosser) [inline test/inductor/test_flex_attention.py:2385]: this is super annoying - can't we drop these nodes? are they left in because we don't know if `is_nested` has side effects or not?

**Review by jbschlosser — COMMENTED** [review]
ping @drisspg / @Chillee this is ready for review when you get the chance :) I booted support for "non-contiguous with holes" NJTs to a future PR

File: test/inductor/test_flex_attention.py (hunk 3 above)
Comment (jbschlosser) [reply test/inductor/test_flex_attention.py:2622]: I wouldn't expect that we'd need guards since `is_nested` is a property of a normal tensor and nothing is being done with the return value, but correct me if I'm wrong.

Comment (jbschlosser) [reply test/inductor/test_flex_attention.py:2622]: As discussed offline, I ran benchmarking before and after this PR
```
python benchmarks/transformer/score_mod.py --decoding
```

Before:
```
| Type    |   Speedup | score_mod     | mask_mod   | dtype          | shape(B,Hq,M,Hkv,N,D)    |
|---------|-----------|---------------|------------|----------------|--------------------------|
| Average |     2.743 |               |            |                |                          |
| Max     |    10.935 | relative_bias | None       | torch.bfloat16 | (16, 16, 1, 2, 4096, 64) |
| Min     |     0.256 | None          | causal     | torch.bfloat16 | (2, 16, 1, 16, 512, 128) |
```

After:
```
| Type    |   Speedup | score_mod     | mask_mod   | dtype          | shape(B,Hq,M,Hkv,N,D)    |
|---------|-----------|---------------|------------|----------------|--------------------------|
| Average |     2.710 |               |            |                |                          |
| Max     |    10.963 | relative_bias | None       | torch.bfloat16 | (16, 16, 1, 2, 4096, 64) |
| Min     |     0.270 | None          | causal     | torch.bfloat16 | (2, 16, 1, 16, 512, 128) |
```

From what I can tell, speedup number differences are within natural variance.

**Review by drisspg — APPROVED** [review]

File: torch/testing/_internal/common_device_type.py [hunk 4]
```diff
@@ -1950,3 +1950,12 @@ def skipPRIVATEUSE1(fn):
 #  This should probably enumerate all available device type test base classes.
 def get_all_device_types() -> List[str]:
     return ["cpu"] if not torch.cuda.is_available() else ["cpu", "cuda"]
+
+
+flex_attention_supported_platform = unittest.skipUnless(
+    torch.cuda.is_available()
+    and torch.version.hip is None
```
[the remark this answers is not in the document]
Comment (jataylo) [reply torch/testing/_internal/common_device_type.py:1957]: cc: @jbschlosser @drisspg

[the remark this answers is not in the document]
Comment (jataylo) [reply torch/testing/_internal/common_device_type.py:1957]: Please can we re-enable ROCm testing for flex here @jbschlosser, if there are UT failures from this change which required skipping on ROCm this should be communicated not just skipping all tests on ROCm. cc: @jeffdaily @jithunnair-amd

[the remark this answers is not in the document]
Comment (drisspg) [reply torch/testing/_internal/common_device_type.py:1957]: Yeah this is a mistake

**Comment by ezyang** [comment]
Were there any performance benchmarks done for this?
